Cargo.toml

```diff
@@ -30,8 +30,8 @@ filetime = "0.2.9"
 flate2 = { version = "1.0.3", default-features = false, features = ["zlib"] }
 git2 = "0.16.0"
 git2-curl = "0.17.0"
-gix = { version = "0.38.0", default-features = false, features = ["blocking-http-transport-curl", "progress-tree"] }
-gix-features-for-configuration-only = { version = "0.27.0", package = "gix-features", features = [ "parallel" ] }
+gix = { version = "0.39.0", default-features = false, features = ["blocking-http-transport-curl", "progress-tree"] }
+gix-features-for-configuration-only = { version = "0.28.0", package = "gix-features", features = [ "parallel" ] }
```

Comment:
Forgot to ask in the previous PR review. Could you give me a pointer to how this configuration works?

Reply:
The `gix-features` crate contains all parts of an implementation that are changeable at compile-time with cargo features. Plumbing crates pull it in and if necessary, enable 'base' versions of the capabilities they need. Those are always pure-Rust implementations of algorithms like `zlib` or `SHA1`. The `gix-features` crate also contains primitives for parallelism and utilities to parallelize work. 

The `gix` crate as entry-point for application developers now pulls in `gix-features` to enable typical features that add on top of the baseline, like enabling actual parallelism with threads, or better performing versions of core algorithms. This works because cargo-feature resolution will compile `gix-features` only once and aggregate all features set by all crates that use it.

Now, `cargo` has disabled default features for `gix` and only re-enables a few ones. None of the features provided by the `gix-features` crate are 'forwarded' in the `gix` crate to avoid duplication, so `cargo` has to pull it in itself to get a chance to set the features it needs. For now, this is only `parallel` to get parallel computation for everything. but it could also enable faster SHA1 or zlib implementations, but that would pull in more C - right now `gitoxide` is configured to be 'pure Rust' which should be most portable.

You can try it yourself by commenting out the `gix-features-for-configuration-only …` line and fetch the crates index - it will be single-threaded then and quite a bit longer. It will still be faster than the `git2` version though as it's optimal, and doesn't do any unnecessary computation. It's quite neat to see that the `gix_feature::parallel` functions also exist in API-compatible serial forms that work the same - those kick in when `parallel` is not set.

I think @epage called it a 'cargo-features-hack', and that's certainly a fitting name as well 😁.

Reply:
Thank you so much! I didn't expect such a comprehensive explanation 😆. Really really useful!